### PR TITLE
Add Swagger UI Express in place of speccy

### LINKED ISSUE #325: Migrate from Speccy to SwaggerUI
Speccy has some vulnerabilities that bug us every time we run `npm install`, also it could be organised better like Swagger for example.

So back in [December](https://github.com/freeCodeCamp/chapter/wiki/Meeting-Notes-December-15,-2019) we decided to us SwaggerUI.

We could probably use something like [SwaggerUIExpress](https://www.npmjs.com/package/swagger-ui-express)

### PR DESCRIPTION
**Pull request by AshNaz87**
- [x] I have read [Chapter's contributing guidelines](https://github.com/freeCodeCamp/chapter/blob/master/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update README.md`).
- [x] My pull request targets the `master` branch of Chapter.

Closes #325

### CODE REVIEW

**Comment by allella** [comment]
@nik-john  you able to review the Swagger?

**Comment by nik-john** [comment]
@allella sure

**Comment by nik-john** [comment]
Second @Zeko369 - a whole lot of code exists already that needs to be migrated. If I understand correctly, `swagger-ui-express` can use the `swagger.json` file

**Comment by Zeko369** [comment]
Yeah, `swagger-ui-express` is just a tool to host swagger/openAPI docs, so our `swagger.json` will work with it no problem

**Comment by Zeko369** [comment]
@AshNaz87 Any progress?
Also I changed dev-dep to dep because you'll use it in a script and `eslint` will yell at you if it's not a dependency

**Comment by AshNaz87** [comment]
@Zeko369 

I'm not sure I fully follow what is required. Care to break it down in steps?

**Comment by Zeko369** [comment]
We use a tool called speccy for hosting our OpenAPI spec (commonly known as Swagger spec). That file is located in `api/swagger.json`.

The job of this PR is to create a script that we'll be able to use instead of speccy.

My suggestion was to use `swagger-ui-expresss`, a lib used in conjunction with express to host a Swagger UI explorer for OpenAPI specs.

You have to options on how to implement it:
1. You could add it as a route to our main app (`server/index.ts`) if NODE_ENV is dev, becase I don't think that we should host it publicly for the MVP.
2. The other option would be a drop in replacement for `speccy`. You could just create a `api/index.ts` file and host an express server with swaggerUiExpress on root route and add that as a script in package.json. Also in this case you would replace the current calls to speccy in `docker-compose.yml`, compared to the first option where you could just remote it.

There a lot of examples in the `swaggerUiExpress` readme file (you can easily find it in the npm page).

If you have any other questions reach out here (preferably) or in discord.

**Comment by Zeko369** [comment]
@AshNaz87 you still working on this?, If not maybe @ceciliaconsta3 could give it a shot?

**Comment by AshNaz87** [comment]
@Zeko369 Apologies for the late reply

Am still working on it... Little busy with work but should submit a PR later this evening (UTC time)

**Comment by Zeko369** [comment]
@AshNaz87 OK, I'll take a look at it in the evening (I'm GMT+1)

**Comment by AshNaz87** [comment]
@Zeko369 Sorry about not getting this done sooner

Let me know about any changes that are required

Thanks

File: server/index.ts [hunk 1]
```diff
@@ -36,4 +38,6 @@ nextjs.nextApp.prepare().then(async () => {
   app.listen(port, () => {
     console.log(`\n\nstarted on port ${port}\n\n`); // tslint:disable-line no-console
   });
+
+  app.use('/api-docs', swaggerUi.serve, swaggerUi.setup(swaggerDocument));
```
Comment (Zeko369) [inline server/index.ts]: I think that we should wrap this in `if nodeenv === prod`, @timmyichen @allella do we want API docs to be public in the MVP

Comment (Zeko369) [reply server/index.ts]: Also maybe change it to `/api/v1/(docs||speccy)` to represent that these are the docs for the `API v1`

File: package.json [hunk 2]
```diff
@@ -62,6 +62,7 @@
     "redux": "^4.0.4",
     "redux-thunk": "^2.3.0",
     "reflect-metadata": "^0.1.13",
+    "swagger-ui-express": "^4.1.3",
```
Comment (Zeko369) [inline package.json]: add [types](https://www.npmjs.com/package/@types/swagger-ui-express)

File: server/index.ts [hunk 3]
```diff
@@ -6,6 +6,8 @@ import { responseErrorHandler } from 'express-response-errors';
 import nextjs from 'server/lib/next';
 import { initDB } from 'server/db';
 import { apiV1 } from './routes';
+import swaggerUi from 'swagger-ui-express';
```
Comment (Zeko369) [inline server/index.ts]: Order imports (first global - npm, then local)

**Comment by AshNaz87** [comment]
@Zeko369 Made changes and pushed up

File: server/index.ts [hunk 4]
```diff
@@ -36,4 +38,8 @@ nextjs.nextApp.prepare().then(async () => {
   app.listen(port, () => {
     console.log(`\n\nstarted on port ${port}\n\n`); // tslint:disable-line no-console
   });
+
+  if (process.env.NODE_ENV === 'production') {
```
Comment (Zeko369) [inline server/index.ts]: Crap, my fault I meant development 🤦‍♂

File: package.json [hunk 5]
```diff
@@ -42,6 +42,7 @@
   "dependencies": {
     "@material-ui/core": "^4.5.2",
     "@material-ui/icons": "^4.5.1",
+    "@types/swagger-ui-express": "^4.1.1",
```
Comment (Zeko369) [inline package.json]: Just move it down to devDependencies

**Comment by AshNaz87** [comment]
@Zeko369 

Pushed up further changes

Think the conflict is with the `package-lock.json` file. Should this be git ignored?

**Comment by vkWeb** [comment]
@Zeko369 Shouldn't `package.json` scripts be updated? Also, we should remove speccy from dev dependency if we haven't already. 

We are still serving speccy:
https://github.com/freeCodeCamp/chapter/blob/260fdc563710e99f4de3cc7f16d77377d78d0901/package.json#L21

**Comment by Zeko369** [comment]
Yes @vkWeb, also docker-compose should be updated

**Comment by vkWeb** [comment]
@Zeko369 Agreed.

**Comment by AshNaz87** [comment]
Not sure how to change the script and the docker-compose files

**Comment by AshNaz87** [comment]
@Zeko369 @vkWeb 

Wouldn't amending package.json and docker-compose.yml create conflicts with master?

**Comment by Zeko369** [comment]
What do you mean by the last comment?

**Comment by AshNaz87** [comment]
Removing `speccy` would create a conflict

**Comment by Zeko369** [comment]
Why?, also we can just merge this and I can create a PR to fix it

**Comment by AshNaz87** [comment]
Might sound really silly, but I'm not sure how to implement this properly

**Comment by vkWeb** [comment]
@AshNaz87 Thanks for your help on this so far. Every little bit counts 😃 

@Zeko369 Let's get this in and we'll land the remaining bits with https://github.com/freeCodeCamp/chapter/pull/339.

**Pull request merged**